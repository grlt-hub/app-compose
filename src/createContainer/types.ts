import { type NonEmptyTuple } from '@shared';
import { type StoreWritable } from 'effector';

type AnyObject = Record<string, unknown>;
type ValueOf<T> = T[keyof T];

const CONTAINER_STATUS = {
  idle: 'idle',
  pending: 'pending',
  done: 'done',
  fail: 'fail',
  off: 'off',
} as const;

type ContainerStatus = ValueOf<typeof CONTAINER_STATUS>;
type StartResult<T> = Promise<{ api: T }> | { api: T };
type EnableResult = Promise<boolean> | boolean;

type AnyAPI = AnyObject | null;
type AnyStartFn = (...x: any) => StartResult<AnyAPI>;

type AnyContainer = {
  id: string;
  domain: string;
  $status: StoreWritable<ContainerStatus>;
  start: AnyStartFn;
  dependsOn?: AnyContainer[];
  optionalDependsOn?: AnyContainer[];
  enable?: (..._: any) => EnableResult;
};

type AnyDeps = NonEmptyTuple<AnyContainer> | void;
type ContainerId = AnyContainer['id'];
type ContainerDomain = AnyContainer['domain'];

export { CONTAINER_STATUS };
export type { AnyAPI, AnyContainer, AnyDeps, ContainerDomain, ContainerId, ContainerStatus, EnableResult, StartResult };
