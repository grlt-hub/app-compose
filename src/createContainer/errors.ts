const ERROR = {
  CONTAINER_ID_EMPTY_STRING: 'Container ID cannot be an empty string.',
} as const;

type ContainerIdEmptyStringError = { id: never; error: typeof ERROR.CONTAINER_ID_EMPTY_STRING };

export { ERROR, type ContainerIdEmptyStringError };
