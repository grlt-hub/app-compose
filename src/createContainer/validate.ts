import { isNil } from '@shared';
import { type AnyContainer } from './types';

type ValidateParams = Pick<AnyContainer, 'id' | 'domain' | 'dependencies' | 'optionalDependencies'>;

const ERROR = {
  CONTAINER_ID_EMPTY_STRING: 'Container ID cannot be an empty string.',
  CONTAINER_DOMAIN_NAME_EMPTY_STRING: 'Container Domain cannot be an empty string.',
  depsIntersection: (intersection: string[], containerId: ValidateParams['id']) =>
    `Dependency conflict detected in container "${containerId}":` +
    '\n' +
    `The following dependencies are listed as both required and optional: [${intersection.join(', ')}].` +
    '\n\n' +
    `Each dependency should be listed only once, as either required or optional.`,
} as const;

type ContainerIdEmptyStringError = ValidateParams & { id: never; error: typeof ERROR.CONTAINER_ID_EMPTY_STRING };
type ContainerDomainNameEmptyStringError = ValidateParams & {
  domain: never;
  error: typeof ERROR.CONTAINER_DOMAIN_NAME_EMPTY_STRING;
};

const validateContainerId = (x: ValidateParams) => {
  if (isNil(x.id) || x.id.length === 0) {
    throw new Error(ERROR.CONTAINER_ID_EMPTY_STRING);
  }
};

const validateDomainName = (x: ValidateParams) => {
  if (isNil(x.domain) || x.domain.length === 0) {
    throw new Error(ERROR.CONTAINER_DOMAIN_NAME_EMPTY_STRING);
  }
};

const validateDepsIntersection = (params: ValidateParams) => {
  if (isNil(params.dependencies) || isNil(params.optionalDependencies)) {
    return;
  }

  const depIds = new Set(params.dependencies.map((dep) => dep.id));
  const optDepsIds = new Set(params.optionalDependencies.map((dep) => dep.id));

  const intersection = depIds.intersection(optDepsIds);

  if (intersection.size) {
    throw new Error(ERROR.depsIntersection(Array.from(intersection), params.id));
  }
};

type InferValidParams<T extends {}> = Exclude<T, ContainerIdEmptyStringError | ContainerDomainNameEmptyStringError>;

const validate = <P extends ValidateParams>(params: P) => {
  validateContainerId(params);
  validateDomainName(params);
  validateDepsIntersection(params);

  return params as InferValidParams<P>;
};

export { validate, type ContainerDomainNameEmptyStringError, type ContainerIdEmptyStringError };
