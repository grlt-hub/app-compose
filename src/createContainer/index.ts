import { createStore } from 'effector';
import type { AnyAPI, AnyContainer, AnyDeps, EnableResult, StartResult, Status } from './types';
import { type ContainerIdEmptyStringError, validate } from './validate';

type ExtractDeps<D extends AnyContainer[]> = {
  [K in D[number] as Awaited<ReturnType<K['start']>>['api'] extends Record<string, never> ? never : K['id']]: Awaited<
    ReturnType<K['start']>
  >['api'];
};

type Params<
  Id extends string,
  API extends AnyAPI,
  Deps extends AnyDeps = void,
  OptionalDeps extends AnyDeps = void,
> = '' extends Id
  ? ContainerIdEmptyStringError
  : Deps extends void
    ? OptionalDeps extends void
      ? {
          id: Id;
          start: () => StartResult<API>;
          enable?: () => EnableResult;
        }
      : {
          id: Id;
          optionalDependsOn: Exclude<OptionalDeps, void>;
          start: (_: void, optionalDeps: Partial<ExtractDeps<Exclude<OptionalDeps, void>>>) => StartResult<API>;
          enable?: (_: void, optionalDeps: Partial<ExtractDeps<Exclude<OptionalDeps, void>>>) => EnableResult;
        }
    : OptionalDeps extends void
      ? {
          id: Id;
          dependsOn: Exclude<Deps, void>;
          start: (deps: ExtractDeps<Exclude<Deps, void>>) => StartResult<API>;
          enable?: (deps: ExtractDeps<Exclude<Deps, void>>) => EnableResult;
        }
      : {
          id: Id;
          dependsOn: Exclude<Deps, void>;
          optionalDependsOn: Exclude<OptionalDeps, void>;
          start: (
            deps: ExtractDeps<Exclude<Deps, void>>,
            optionalDeps: Partial<ExtractDeps<Exclude<OptionalDeps, void>>>,
          ) => StartResult<API>;
          enable?: (
            deps: ExtractDeps<Exclude<Deps, void>>,
            optionalDeps: Partial<ExtractDeps<Exclude<OptionalDeps, void>>>,
          ) => EnableResult;
        };

const createContainer = <
  Id extends string,
  API extends AnyAPI,
  Deps extends AnyDeps = void,
  OptionalDeps extends AnyDeps = void,
>(
  __params: Params<Id, API, Deps, OptionalDeps>,
) => {
  const params = validate(__params);
  const $status = createStore<Status>('idle');

  return {
    ...params,
    $status,
  };
};

export { createContainer, type AnyContainer };
