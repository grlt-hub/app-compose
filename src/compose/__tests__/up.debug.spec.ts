import { CONTAINER_STATUS, createContainer, type AnyContainer } from '../../createContainer';
import { compose } from '../index';

test('up.debug', async () => {
  const consoleLogSpy = vi.spyOn(console, 'debug').mockImplementation(() => {});

  const a = createContainer({ id: 'a', start: () => ({ api: null }) });
  const b = createContainer({ id: 'b', dependsOn: [a], enable: () => false, start: () => ({ api: null }) });
  const c = createContainer({ id: 'c', start: () => ({ api: null }) });

  await compose.up([a, b, c], { debug: true });

  expect(consoleLogSpy.mock.calls[0]).toStrictEqual([{ a: 'idle', b: 'idle', c: 'idle' }]);
  expect(consoleLogSpy.mock.calls[1]).toStrictEqual([{ a: 'pending', b: 'idle', c: 'idle' }]);
  expect(consoleLogSpy.mock.calls[2]).toStrictEqual([{ a: 'pending', b: 'idle', c: 'pending' }]);
  expect(consoleLogSpy.mock.calls[3]).toStrictEqual([{ a: 'done', b: 'idle', c: 'pending' }]);
  expect(consoleLogSpy.mock.calls[4]).toStrictEqual([{ a: 'done', b: 'idle', c: 'done' }]);
  expect(consoleLogSpy.mock.calls[5]).toStrictEqual([{ a: 'done', b: 'off', c: 'done' }]);

  consoleLogSpy.mockRestore();
});
