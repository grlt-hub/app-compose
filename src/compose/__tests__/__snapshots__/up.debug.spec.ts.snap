// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`up.debug 1`] = `
[
  "[2024-01-01T00:00:00.000Z] app-compose:",
  "{
  "a": "idle",
  "b": "idle",
  "c": "idle"
}",
]
`;

exports[`up.debug 2`] = `
[
  "[2024-01-01T00:00:00.000Z] app-compose:",
  "{
  "a": "pending",
  "b": "idle",
  "c": "idle"
}",
]
`;

exports[`up.debug 3`] = `
[
  "[2024-01-01T00:00:00.000Z] app-compose:",
  "{
  "a": "pending",
  "b": "idle",
  "c": "pending"
}",
]
`;

exports[`up.debug 4`] = `
[
  "[2024-01-01T00:00:00.000Z] app-compose:",
  "{
  "a": "done",
  "b": "idle",
  "c": "pending"
}",
]
`;

exports[`up.debug 5`] = `
[
  "[2024-01-01T00:00:00.000Z] app-compose:",
  "{
  "a": "done",
  "b": "idle",
  "c": "done"
}",
]
`;

exports[`up.debug 6`] = `
[
  "[2024-01-01T00:00:00.000Z] app-compose:",
  "{
  "a": "done",
  "b": "off",
  "c": "done"
}",
]
`;
